nexus/db-queries/src/db/queries/next_item.rs

```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+///     -- in the table.
+///     (
+///         SELECT next_item AS <item_column>
+///         FROM (
+///             SELECT
+///                 <item_column> + 1 AS next_item
+///             FROM
+///                 <table>
+///             WHERE
+///                 <scope_column> = <scope_key> AND
+///                 time_deleted IS NULL
+///         )
+///         LEFT OUTER JOIN
+///             <table>
+///         ON
+///             (<scope_column>, next_item <= <max_item>, time_deleted IS NULL) =
+///             (<scope_key>, TRUE, TRUE)
+///         WHERE
+///             <item_column> IS NULL AND next_item <= <max_item>
+///         ORDER BY next_item
+///         LIMIT 1
+///     ),
+///     -- If the table _is_ empty, just insert the minimum value.
+///     <min_item>
+/// )
+/// ```
+///
+/// # Which one to use?
+///
+/// One should probably prefer to use this form of the query, if possible. Both
+/// queries appear to read roughly the same amount of data from disk, but the
+/// series-based implementation uses far more memory and runs more slowly.
+///
+/// One downside of this query is that it always allocates the next _smallest_
+/// item that's available. The `NextItem` query lets the caller choose a base
+/// from which to start the search, which is useful in situations where one
+/// would prefer to randomly distribute items rather than sequentially allocate
+/// them.
+#[derive(Debug, Clone, Copy)]
+pub(super) struct NextItemSelfJoined<
+    Table,
+    Item,
+    ItemColumn,
+    ScopeKey = NoScopeKey,
+    ScopeColumn = NoScopeColumn,
+> {
+    table: Table,
+    _d: PhantomData<(Item, ItemColumn, ScopeColumn)>,
+    scope_key: ScopeKey,
+    item_min: Item,
+    item_max: Item,
+}
+
+impl<Table, Item, ItemColumn, ScopeKey, ScopeColumn>
+    NextItemSelfJoined<Table, Item, ItemColumn, ScopeKey, ScopeColumn>
+where
+    // Table is a database table whose name can be used in a query fragment
+    Table: diesel::Table + HasTable<Table = Table> + QueryFragment<Pg> + Copy,
+
+    // Item can be converted to the SQL type of the ItemColumn
+    Item: ToSql<<ItemColumn as Expression>::SqlType, Pg> + Copy,
+
+    // ItemColum is a column in the target table
+    ItemColumn: Column<Table = Table> + Copy,
+
+    // ScopeKey can be converted to the SQL type of the ScopeColumn
+    ScopeKey: ScopeKeyType + ToSql<<ScopeColumn as Expression>::SqlType, Pg>,
+
+    // ScopeColumn is a column on the target table
+    ScopeColumn: ScopeColumnType + Column<Table = Table>,
+
+    // The Postgres backend supports the SQL types of both columns
+    Pg: HasSqlType<<ScopeColumn as Expression>::SqlType>
+        + HasSqlType<<ItemColumn as Expression>::SqlType>,
+{
+    /// Create a new `NextItemSelfJoined` query, scoped to a particular key.
+    pub(super) fn new_scoped(
+        scope_key: ScopeKey,
+        item_min: Item,
+        item_max: Item,
+    ) -> Self {
```

Comment:
The intention is that these are both inclusive, right? Could we document that?

Reply:
Documented in 75311d5a7